---
title: "Genome-wide association analyses implemented using the GENESIS software"
output: html_document
date: "08-02-2023"
---

# Script 

```{r}
# Install GENESIS package 
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("GENESIS")
#Install gdsfmt package
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("gdsfmt")
```

```{r}
# Library required packages
suppressMessages(library(GENESIS))
suppressMessages(library(GWASTools))
suppressMessages(library(SNPRelate))
suppressMessages(library(data.table))
suppressMessages(library(dplyr))
library(GENESIS)
library(GWASTools)
library(SNPRelate)
library(gdsfmt)
library(qqman)
```

## Editing working directory
Directory to be edited with paths to relevant input files.
Please edit the following paths and names:
"phenotye.csv"
"cov.csv"
"pc.csv"

```{r}
#Change working directory to paths to relevant input files: Phenotype, Covariates, PC scores
dir_con <- c("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/results/data_harmonization/UKB/outputs/do_not_share/updated_phenotypes") #Directory to DINT-transformed continuous phenotypes of interest
dir_bi <- c("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/data/data_harmonization/UKB") #Directory to binary phenotype of interest input 
dir <- c("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/data/GWAS")
diro <- c("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/results/GWAS") #e.g. "your-home-directory/bp-drug-mr/outputs"
phen_con <- fread(paste0(dir_con, "/EAS/mrc_phenotypes_eas_int.txt"), header=T, fill = T, na.strings = c("", NA), data.table = F, stringsAsFactors = FALSE)
phen_bi <- fread(paste0(dir_bi, "mrc_phenotypes_eas.txt"), header=T, fill = T, na.strings = c("", NA), data.table = F, stringsAsFactors = FALSE)
cov <- fread(paste0("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/data/data_harmonization/UKB/mrc_covariates.txt"), header=T, fill = T, na.strings = c("", NA), data.table = F, stringsAsFactors = FALSE)
pc <- fread(paste0("/projects/MRC-IEU/research/projects/ieu2/p4/080/working/data/data_harmonization/UKB/mrc_pca.txt"), header=T, fill = T, na.strings = c("", NA), data.table = F, stringsAsFactors = FALSE)
```


## Join phenotype, covariate and PC data files and prepare input data
```{r}
phen_name_con <- names(phen_con)[-1] 
phen_name_bi <- names(phen_bi)[-1] 
cont <- phen_name_con[phen_con %in% c("sbp", "dbp", "pp", "tc", "ldlc", "tg")]
bi <- phen_name_bi[phen_bi %in% c("chd", "stroke", "hypertension")]
cov_select <- cov %>% select(IID, age, sex) #if analysing lipid traits include med_ldlc
cov_replace <- cov_select %>% mutate(sex = ifelse(sex == 1,"M","F"))
for (i in cont)
  {
    phen_select_con <- phen %>% select("IID", i)
    phe_select_con <- subset(phe_select_con, !is.na(phe_select_con[[i]]))
  
   if(i=="sbp"){phe_select_con <- subset(phe_select_con, !(IID %in% out$sbp))}
   if(i=="dbp"){phe_select_con <- subset(phe_select_con, !(IID %in% out$dbp))}
   if(i=="ldlc"){phe_select_con <- subset(phe_select_con, !(IID %in% out$ldlc))}
   dat <- merge(phen_select_con, cov_replace, by="IID") %>% merge(., pc, by="IID") 
    
   dat <- dat %>% na.omit()
    
  if(i=="sbp" | i=="dbp" | i=="pp")
   input_dat_bp <- data.frame(scanID = dat$ieu, pheno = dat[[i]], age = dat$age, sex = dat$sex, pc1 = dat$PC1, pc2 =          dat$PC2, pc3 = dat$PC3, pc4 = dat$PC4, pc5 = dat$PC5, pc6 = dat$PC6, pc7 = dat$PC7, pc8 = dat$PC8, pc9 = dat$PC9, pc10=    dat$PC10, pc11 = dat$PC11, pc12 = dat$PC12, pc13 = dat$PC13, pc14 = dat$PC14, pc15 = dat$PC15, pc16 = dat$PC16, pc17 =      dat$PC17, pc18 = dat$PC18, pc19 = dat$PC19, pc20 = dat$PC20)
    scanAnnot_bp <- ScanAnnotationDataFrame(input_dat_bp)
    scanAnnot_bp
  }
    
  if(i=="ldlc" | i=="tg" | i=="tc") 
    input_dat_lipid <- data.frame(scanID = dat$ieu, pheno = dat[[i]], med_ldlc = dat$med_ldlc, age = dat$age, sex =             dat$sex, pc1= dat$PC1, pc2 = dat$PC2, pc3 = dat$PC3, pc4 = dat$PC4, pc5 = dat$PC5, pc6 = dat$PC6, pc7 = dat$PC7,           pc8= dat$PC8, pc9 =dat$PC9, pc10 = dat$PC10, pc11 = dat$PC11, pc12 = dat$PC12, pc13 = dat$PC13, pc14 = dat$PC14,           pc15 = dat$PC15, pc16 = dat$PC16, pc17 = dat$PC17, pc18 = dat$PC18, pc19 = dat$PC19, pc20 = dat$PC20)
    
    scanAnnot_lipid <- ScanAnnotationDataFrame(input_dat_lipid)
    scanAnnot_lipid
  } 
  
phen_name_bi <- names(phen_bi)[-1] 
bi <- phen_name_bi[phen_bi %in% c("chd", "stroke", "hypertension")]
cov_select <- cov %>% select(IID, age, sex) #if analysing lipid traits include med_ldlc
cov_replace <- cov_select %>% mutate(sex = ifelse(sex == 1,"M","F"))
for (i in bi)
  {
    phen_select_bi <- phen %>% select("IID", i)
    phe_select_bi <- subset(phe_select_bi, !is.na(phe_select_bi[[i]]))
  
    dat <- merge(phen_select_bi, cov_replace, by="IID") %>% merge(., pc, by="IID") 
    dat <- dat %>% na.omit()
    
    if(i=="chd" | i=="stroke" | i=="hypertension") 
    input_dat_bi <- data.frame(scanID = dat$ieu, pheno = dat[[i]], med_ldlc = dat$med_ldlc, age = dat$age, sex =               dat$sex, pc1= dat$PC1, pc2 = dat$PC2, pc3 = dat$PC3, pc4 = dat$PC4, pc5 = dat$PC5, pc6 = dat$PC6, pc7 = dat$PC7,           pc8= dat$PC8, pc9 =dat$PC9, pc10 = dat$PC10, pc11 = dat$PC11, pc12 = dat$PC12, pc13 = dat$PC13, pc14 = dat$PC14,           pc15 = dat$PC15, pc16 = dat$PC16, pc17 = dat$PC17, pc18 = dat$PC18, pc19 = dat$PC19, pc20 = dat$PC20)
    
    scanAnnot_bi <- ScanAnnotationDataFrame(input_dat_bi)
    scanAnnot_bi
  } 
```

## Read in genotype data 
```{r}
snpgdsBED2GDS(bed.fn = paste0(dir, "/UKB/GENESIS/chr1-22_merged.EAS.bed"),
              bim.fn = paste0(dir, "/UKB/GENESIS/chr1-22_merged.EAS.bim"),
              fam.fn = paste0(dir, "/UKB/GENESIS/chr1-22_merged.EAS.fam"),
              out.gdsfn = paste0(dir, "/UKB/GENESIS/EAS.genotype.gds")
              
geno <- GdsGenotypeReader(filename = paste0(dir, "/UKB/GENESIS/EAS.genotype.gds"))
genoData <- GenotypeData(geno)
showfile.gds(closeall=TRUE)
```
## Principal components analysis in related samples (PC-AiR) 
```{r}
# set seed for LD pruning
set.seed(100)
# LD pruning
genoData <- snpgdsOpen(filename = paste0(dir, "/UKB/GENESIS/EAS.genotype.gds"))
snpset <- snpgdsLDpruning(genoData, method="corr", slide.max.bp=10e6, ld.threshold=sqrt(0.1), verbose=FALSE)
pruned <- unlist(snpset, use.names=FALSE)
# Pairwise measures of ancestry divergence 
KINGmat <- snpgdsIBDKING(genoData, sample.id=NULL, snp.id=NULL, autosome.only=TRUE,
    remove.monosnp=TRUE, maf=NaN, missing.rate=NaN,
    type=c("KING-robust", "KING-homo"), family.id=NULL, num.thread=1L,
    useMatrix=FALSE, verbose=TRUE)
snpgdsClose(genoData)
# run PC-AiR
geno <- GdsGenotypeReader(filename = paste0(dir, "/UKB/GENESIS/EAS.genotype.gds"))
genoData <- GenotypeData(geno)
mypcair <-pcair(genoData, divobj = KINGmat, snp.include = pruned)
summary(mypcair)
#plot PC-AiR PCs 
png(file= paste0(diro, "/UKB/EAS/GENESIS_pcair_pc1_pc2.png"), width=8, height=7, units="in", res=300)
plot(mypcair)
dev.off() 
png(file= paste0(diro, "/UKB/EAS/GENESIS_pcair_pc3_pc4.png"), width=8, height=7, units="in", res=300)
plot(mypcair, vx = 3, vy = 4)
dev.off() 
png(file= paste0(diro, "/UKB/EAS/GENESIS_pcair_pc5_pc6.png"), width=8, height=7, units="in", res=300)
plot(mypcair, vx = 5, vy = 6)
dev.off() 
png(file= paste0(diro, "/UKB/EAS/GENESIS_pcair_pc7_pc8.png"), width=8, height=7, units="in", res=300)
plot(mypcair, vx = 7, vy = 8)
dev.off() 
png(file= paste0(diro, "/UKB/EAS/GENESIS_pcair_pc9_pc10.png"), width=8, height=7, units="in", res=300)
plot(mypcair, vx = 9, vy = 10)
dev.off() 
```

## Relatedness estimation adjusted for principal components (PC-Relate)
```{r}
# run PC-Relate
genoData_pruned <- GenotypeBlockIterator(genoData, snpInclude=pruned)
mypcrelate <- pcrelate(genoData_pruned, pcs = mypcair$vectors[,1:20], 
                       training.set = mypcair$unrels)
# plot PC-Relate 
png(file= paste0(diro, "/UKB/EAS/pc_relate.png"), width=8, height=7, units="in", res=300)
plot(mypcrelate$kinBtwn$k0, mypcrelate$kinBtwn$kin, xlab="k0", ylab="kinship")
dev.off() 
# save pc-relate output 
mat2gds(mypcrelate, paste0(dir, "/UKB/GENESIS/EAS.pc.relate.gds"))
# Make a genetic relationship matrix (GRM) from kinship estimates calculated from PC-Relate analysis
myGRM <- pcrelateToMatrix(mypcrelate, thresh = 2^(-11/2), scaleKin = 2)
 ```

## Mixed model association testing 
```{r}
# Fit the null model - Blood pressure traits 
nullmod <- fitNullModel(scanAnnot_bp[[i]], outcome = "pheno", covars = c("age", "sex", "pc1", "pc2", "pc3", "pc4", "pc5", "pc6", "pc7", "pc8", "pc9", "pc10", "pc11", "pc12", "pc13", "pc14", "pc15", "pc16", "pc17", "pc18", "pc19", "pc20"), 
                        cov.mat = myGRM, family = gaussian)
                    
# Run SNP-phenotype association test - Blood pressure traits
genoIterator <- GenotypeBlockIterator(genoData)
assoc <- assocTestSingle(genoIterator, null.model = nullmod, test="Score")
ea <- effectAllele(genoData, variant.id=assoc$variant.id)
assoc$chr <- as.numeric(assoc$chr)
assoc_out <- merge(assoc, ea, by="variant.id")
assoc_out_sorted <- assoc_out[order(assoc_out$chr),]

# Save output files 
write.table(assoc_out_sorted, file=paste0(diro, "/UKB/EAS/GENESIS_assoc_out.txt"), row.names=F, quote=F)

#Create Manhattan and Q-Q plot 
png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_manhattan.png"), width=8, height=7, units="in", res=300)
manhattan(assoc_out_sorted, chr="chr", bp="pos", snp="variant.id", p="Score.pval" )
dev.off()
png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_qq.png"), width=8, height=7, units="in", res=300)
qq(assoc_out_sorted$Score.pval)
dev.off()

# Fit the null model - Lipid traits 
nullmod <- fitNullModel(scanAnnot_lipid[[i]], outcome = "pheno", covars = c("age", "sex", "pc1", "pc2", "pc3", "pc4", "pc5", "pc6", "pc7", "pc8", "pc9", "pc10", "pc11", "pc12", "pc13", "pc14", "pc15", "pc16", "pc17", "pc18", "pc19", "pc20"), 
                        cov.mat = myGRM, family = gaussian)
                    
# Run SNP-phenotype association test - Lipid traits
genoIterator <- GenotypeBlockIterator(genoData)
assoc <- assocTestSingle(genoIterator, null.model = nullmod, test="Score", GxE = scanAnnot$med_ldlc)
ea <- effectAllele(genoData, variant.id=assoc$variant.id)
assoc$chr <- as.numeric(assoc$chr)
assoc_out <- merge(assoc, ea, by="variant.id")
assoc_out_sorted <- assoc_out[order(assoc_out$chr),]

# Save output files 
write.table(assoc_out_sorted, file=paste0(diro, "/UKB/EAS/GENESIS_assoc_out.txt"), row.names=F, quote=F)

#Create Manhattan and Q-Q plot 
png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_manhattan.png"), width=8, height=7, units="in", res=300)
manhattan(assoc_out_sorted, chr="chr", bp="pos", snp="variant.id", p="Score.pval" )
dev.off()

png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_qq.png"), width=8, height=7, units="in", res=300)
qq(assoc_out_sorted$Score.pval)
dev.off()

# Fit the null model - Binary traits 
nullmod <- fitNullModel(scanAnnot_bi[[i]], outcome = "pheno", covars = c("age", "sex", "pc1", "pc2", "pc3", "pc4", "pc5", "pc6", "pc7", "pc8", "pc9", "pc10", "pc11", "pc12", "pc13", "pc14", "pc15", "pc16", "pc17", "pc18", "pc19", "pc20"), 
                        cov.mat = myGRM, family = binomial)
                        
# Run SNP-phenotype association test - binary traits
genoIterator <- GenotypeBlockIterator(genoData)
assoc <- assocTestSingle(genoIterator, null.model = nullmod, test=c("Score", "Score.SPA"))
ea <- effectAllele(genoData, variant.id=assoc$variant.id)
assoc$chr <- as.numeric(assoc$chr)
assoc_out <- merge(assoc, ea, by="variant.id")
assoc_out_sorted <- assoc_out[order(assoc_out$chr),]

# Save output files 
write.table(assoc_out_sorted, file=paste0(diro, "/UKB/EAS/GENESIS_assoc_out.txt"), row.names=F, quote=F)

#Create Manhattan and Q-Q plot 
png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_manhattan.png"), width=8, height=7, units="in", res=300)
manhattan(assoc_out_sorted, chr="chr", bp="pos", snp="variant.id", p="Score.pval" )
dev.off()

png(file= paste0(diro, "/UKB/EAS/GENESIS_sbp_assoc_qq.png"), width=8, height=7, units="in", res=300)
qq(assoc_out_sorted$Score.pval)
dev.off()
